Unload dependents of a mod which failed to load

MacaylaMarvelous81 opened the pull request:
### What is the bug?
#3821 - If a mod which is a hard dependency of other mods is unloaded due to a load error, the dependency mod is unloaded while the dependent mod(s) are not unloaded. This can cause a load loop since the culprit mod will be loaded again when dependencies are automatically enabled.

### How did you fix the bug?
With this pull request, when a mod is unloaded because of a load error, the mod loader will search for mods which depend on it and disable them. It does this by checking if any of the `availableMods` have the dependency mod in the `modReferences`.

### Are there alternatives to your fix?
I'm not aware of any other way to get the dependents of a mod other than doing the inverse and searching the dependencies of other mods to find them. Another way to get the dependents would likely be more logical for reading the code, and it would be better if not every mod would have to be checked.

I'm also unsure about the message localization. I added a string to the en-US localization so that the message will indicate which dependents were disabled, if any, but I'm unfamiliar with how localization works in tModLoader. Is it OK for me to only add it to the en-US file?

MacaylaMarvelous81 committed:
Unload dependents when a hard dependency fails to load

MacaylaMarvelous81 committed:
Include disabled dependent mods in message

Mods that were disabled because a hard dependency encountered a load error will also appear in the error message as they are being unloaded.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModLoader.cs:
```diff
@@ -176,11 +176,22 @@ private static void Load(CancellationToken token = default)
+			foreach (var mod in responsibleMods) {
 				DisableMod(mod);
 
+				var dependents = from m in availableMods
```
We don't use sql linq syntax in tML. Please switch to functional style.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModLoader.cs:
```diff
@@ -176,11 +176,22 @@ private static void Load(CancellationToken token = default)
 				DisableMod(mod);
 
+				var dependents = from m in availableMods
+					where m.properties.modReferences.Any(reference => reference.mod.Equals(mod))
```
`BuildProperties` has the `RefNames` method which will help here. We used explicit named args when passing constants, so you'll want to do `RefNames(includeWeak: false)` which makes it really clear that you're only considering strong references as required deps.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModLoader.cs:
```diff
@@ -176,11 +176,22 @@ private static void Load(CancellationToken token = default)
+					where m.properties.modReferences.Any(reference => reference.mod.Equals(mod))
+					select m.Name;
+
+				msg += "\n" + Language.GetTextValue("tModLoader.LoadErrorDependentsDisabled", mod,
```
Please aggregate all the disabled dependents into one list so the message only gets added once. 

Please also filter to only enabled mods.

Additionally, you will need to write this in a recursive manner, because you may have a series of mod references `B requires C requires A` and if `A` gets disabled, then you have to disable `C` and then `B` as well.

MacaylaMarvelous81 committed:
Use functional-style LINQ query

This commit changes the LINQ query for finding mods which depend on another mod to adhere to tModLoader code style by using functions rather than the query syntax.

Relevant discussion: https://github.com/tModLoader/tModLoader/pull/3829#discussion_r1337892779

MacaylaMarvelous81 committed:
Use RefNames to find mod references

Changes the dependent mods query to use BuildProperties.RefNames rather than looking through BuildProperties.modReferences.

Relevant discussion: https://github.com/tModLoader/tModLoader/pull/3829#discussion_r1337893174

JavidPack commented:
@MacaylaMarvelous81 is this still supposed to be a draft? You seem to have made the changes addressing the feedback already. I'd be happy to finish this up for you if there is anything still left to implement. 

I noticed, however, that the "Allow edits from maintainers." checkbox is not checked for this PR. You'll need to check this if you want me to finish this up, otherwise I can't commit my edits.

![image](https://github.com/tModLoader/tModLoader/assets/4522492/4e19983d-744c-41af-8187-bab1b849f0cd)
